### PR TITLE
[IOCOM-500] Support for cancelled SEND message

### PR DESCRIPTION
**Pull request by Vangaorth**
⚠️ This PR depends on #294 ⚠️

## Short description
This PR adds support for cancelled PN notification messages, with proper timeline event creation.

## List of changes proposed in this pull request
- `src/features/pn/payloads/messages`: added status property to PN message and proper timeline computation to include both paid events and cancelled status (if needed)
- `src/features/pn/types/notificationStatusHistoryElement`: type for PN timeline elements
- `src/utils/payment`: added rptId computation method from NotificationPaymentInfo instance. Added isPaid method to check when a payment is paid

## How to test
Run the server instance with a template configuration that cancels the notification. Retrieve it using the `third-party-messages/:id` endpoint and check both the `status` property and the related timeline events

### CODE REVIEW

File: src/features/pn/payloads/messages.ts [hunk 1]
```diff
@@ -308,34 +323,148 @@ const createPNMessageRecipient = (
...
+            fold(
+              processedPayment =>
+                isPaid(processedPayment.status)
+                  ? O.some(notificationRecipient.payment.noticeCode)
+                  : O.none,
+              _ => O.none
+            )
+          )
+        ),
+        O.flatten
```
Comment (adelloste) [inline src/features/pn/payloads/messages.ts:467]:
```suggestion
        O.chain(paymentStatus =>
          pipe(
            paymentStatus,
            fold(
              processedPayment =>
                isPaid(processedPayment.status)
                  ? O.some(notificationRecipient.payment.noticeCode)
                  : O.none,
              _ => O.none
            )
          )
        )
```

File: src/features/pn/payloads/messages.ts [hunk 2]
```diff
@@ -308,34 +323,148 @@ const createPNMessageRecipient = (
...
+            activeFrom: pipe(
+              new Date(Date.parse(utcStartDateString) + index * 60000),
+              utcDate => `${utcDate.toISOString().slice(0, -1)}+00:00`
+            ),
+            relatedTimelineElements: [
+              "TPAK-PJUT-RALE-202207-X-2_notification_payment_paid_0",
+              `TPAK-PJUT-RALE-202207-X-2_notification_payment_paid_${paidNoticeCodes[index]}`
+            ]
+          } as NotificationStatusHistoryElement)
+      )
```
Comment (adelloste) [inline src/features/pn/payloads/messages.ts:442]:
```suggestion
      ...A.makeBy<NotificationStatusHistoryElement>(
        paidNoticeCodes.length,
        index =>
          ({
            status,
            activeFrom: pipe(
              new Date(Date.parse(utcStartDateString) + index * 60000),
              utcDate => `${utcDate.toISOString().slice(0, -1)}+00:00`
            ),
            relatedTimelineElements: [
              "TPAK-PJUT-RALE-202207-X-2_notification_payment_paid_0",
              `TPAK-PJUT-RALE-202207-X-2_notification_payment_paid_${paidNoticeCodes[index]}`
            ]
          })
      )
```

File: src/features/pn/payloads/messages.ts (hunk 1 above)
Comment (Vangaorth) [reply src/features/pn/payloads/messages.ts]: Done it. I also added and explicit return type to the function's signature in order for the Intellisense to properly detect the type

File: src/features/pn/payloads/messages.ts (hunk 2 above)
Comment (Vangaorth) [reply src/features/pn/payloads/messages.ts]: As a matter of facts, the explicit type is not even needed here, so I removed it

**Review by adelloste — APPROVED** [review]

**Pull request merged**